### DISCUSSION TITLE
failure to join a trusted cluster

### DISCUSSION THREAD

**Discussion by alanlubold** [gravitational/teleport#20301] [Q&A] [answered]
Issue has come up on trying to join a leaf cluster to the root cluster I've been working on.

The specific error is this:

arc@6-0-0-0-12:/etc $ sudo /usr/local/bin2/tctl create /etc/trusted_cluster.yaml
ERROR: access denied to perform action "read" on "trusted_cluster"

(Line 235 in the attached)

I have attached various files to show the config I have on the two machines. My goal here is to add a few more machines to the leaf cluster; they are machines at a remote customer site and the root machine is in the cloud, giving access to the customer machines behind a firewall.

The root machine has the old-style config where the trusted_cluster value is in the teleport.yaml, I tried it multiple times using this value in the trusted_cluster file on the leaf but got the same error (but that's why that value is still in the file on the root). I then went to the ephemeral token version, but had the same error so I thought this wasn't the problem.

I'm guessing I have something wrong in my roles giving me a permissions problem, but I can't se what it is.

Let me know if there is any other data I can provide. Thanks.

[trusted_cluster_issue.txt](https://github.com/gravitational/teleport/files/10418437/trusted_cluster_issue.txt)

**Comment by webvictim** [comment]
Run this instead:

`sudo /usr/local/bin2/tctl --config /etc/teleport2.yaml create /etc/trusted_cluster.yaml `

`tctl` reads `/etc/teleport.yaml` to get its config by default, which is telling it to look at `/var/lib/teleport` - but that's where the backend for `teleport1` lives.
You need `teleport2`'s backend, which is referred to by the `teleport.data_dir` parameter in `/etc/teleport2.yaml` and lives in `/var/lib/teleport2`.

**Comment by alanlubold** [comment]
Did so, thank you for the quick response.  

Got this:

arc@6-0-0-0-12:/etc $ sudo /usr/local/bin2/tctl create /etc/trusted_cluster.yaml --config=/etc/teleport2.yaml
trusted cluster "main2" has been created
arc@6-0-0-0-12:/etc $ 

So good. However, got this:

arc@6-0-0-0-12:/etc $ sudo /usr/local/bin2/tsh clusters
Cluster Name Status  Cluster Type Labels Selected 
------------ ------- ------------ ------ -------- 
main2        online  root                *        
6.0.0.0.12   offline leaf                         

arc@6-0-0-0-12:/etc $

The leaf cluster is supposed to be named 8.1.1.2.1, not 6.0.0.0.12. The latter is the Teleport name for the host (with the dashes replaced with dots). 

And, as you can see, it's offline, andI haven't been able to get it online. 

I have arrived at this point before, managed to remove the broken leaf cluster, removed the var/lib/teleport from the root server and started over, but got one or the other of the errors I've been  seeing.

**Reply by webvictim** [reply]
I think the fact that you're running what's essentially two separate clusters on the same machine is causing a lot of confusion here. I would get it into the state where you have one Teleport process running and you can successfully join that leaf back to your root - then try adding the second process.

**Reply by webvictim** [reply]
It's also worth noting that the `cluster_name` is encoded into the CA generated for the cluster itself and you can't change it after creation. I would be tempted to shut the process down, clear out `/var/lib/teleport2` and restart it again just to make sure it's generating a fresh CA with the correct `cluster_name` encoded.

Also, if you're looking to run an all-in-one Teleport auth/proxy/node you can remove the `auth_servers` and `auth_token` entries from your config file - it'll all connect back over `localhost` automatically anyway. If you're running two Teleport processes you're likely going to need to change the `listen_addr` for `auth_service` and `proxy_service` as they'll conflict with the other Teleport process you're running on the same machine...

**Comment by alanlubold** [comment]
Additional note:

On this last try, I had changed the trusted_cluster.yaml file in the role section, to be this:


  role_map:
  - remote: "admin"
    local: ["arc"]

Originally the remote value was "arc"

Not sure if this has any bearing, but just in case.

**Comment by alanlubold** [comment]
Have started from scratch, and have progressed, but still keep coming up with this:

arc@6-0-0-0-12:/etc $ sudo /usr/local/bin2/tsh clusters --proxy=tel11.arctevity-cloud.com --user=alan.lubold@arctevity.com
Enter password for Teleport user alan.lubold@arctevity.com:
Enter your OTP token:
Cluster Name Status  Cluster Type Labels Selected 
------------ ------- ------------ ------ -------- 
main2        online  root                *        
8.1.1.2.1    offline leaf                         

arc@6-0-0-0-12:/etc $


I can't figure out how to get the leaf cluster to be online. 

Thanks

**Reply by webvictim** [reply]
Are there any errors in the Teleport logs on the root or leaf cluster?

What does `tctl get trusted_cluster` say?

**Comment by alanlubold** [comment]
See attached. Let me know if there are other log files you'd like me to get. Appreciate the help. 

Wasn't entirely sure where you wanted me to run get trusted_cluster, but you can see the results in the attached. I'm still getting this:

arc@6-0-0-0-12:/etc $ 
arc@6-0-0-0-12:/etc $ sudo /usr/local/bin2/tsh clusters
Cluster Name Status  Cluster Type Labels Selected 
------------ ------- ------------ ------ -------- 
main2        online  root                *        
8.1.1.2.1    offline leaf                         

arc@6-0-0-0-12:/etc $ 

And here, on the root, the same:

ubuntu@tel11:/etc$ sudo tsh clusters
Cluster Name Status  Cluster Type Labels Selected 
------------ ------- ------------ ------ -------- 
main2        online  root                *        
8.1.1.2.1    offline leaf                         

ubuntu@tel11:/etc$ 


Thanks.




[tctl_get_trusted_cluster.txt](https://github.com/gravitational/teleport/files/10441518/tctl_get_trusted_cluster.txt)

**Comment by alanlubold** [comment]
Question: the documentation for trusted clusters has several places where logging into one cluster or another is specified. In my setup of these systems, I have ordinary SSH access to both all the time, so are these logons necessary?

**Comment by alanlubold** [comment]
This log entry repeats many times in the leaf cluster:

{"ei":0,"event":"cert.create","uid":"e683e7b5-f448-4807-8b27-2ab2e6794efc","code":"TC000I","time":"2023-01-18T10:00:55.22Z","cluster_name":"8.1.1.2.1","cert_type":"user","identity":{"user":"alan.lubold@arctevity.com","roles":["arc"],"logins":["arc","-teleport-internal-join"],"expires":"2023-01-18T10:01:15.592633949Z","route_to_cluster":"8.1.1.2.1","traits":{"aws_role_arns":null,"db_names":null,"db_users":null,"kubernetes_groups":null,"kubernetes_users":null,"logins":null,"windows_logins":null},"teleport_cluster":"8.1.1.2.1","prev_identity_expires":"0001-01-01T00:00:00Z"}}

On the root server, this entry repeats a lot:

{"ei":0,"event":"cert.create","uid":"cabacc26-cde1-4873-9db9-43a6c87e553c","code":"TC000I","time":"2023-01-18T09:28:35.529Z","cluster_name":"main2","cert_type":"user","identity":{"user":"alan.lubold@arctevity.com","roles":["admin"],"logins":["ubuntu","arc","-teleport-internal-join"],"expires":"2023-01-18T09:28:48.854139503Z","route_to_cluster":"main2","traits":{"aws_role_arns":null,"db_names":null,"db_users":null,"kubernetes_groups":null,"kubernetes_users":null,"logins":null,"windows_logins":null},"teleport_cluster":"main2","prev_identity_expires":"0001-01-01T00:00:00Z"}}

**Comment by alanlubold** [comment]
New leaf cluster from scratch, no odd port numbers or other Teleport installations. Create trusted_cluster command completed affirmative. On leaf machine:

```
arc@remote1:/etc $ sudo tctl get trusted_cluster
kind: trusted_cluster
metadata:
  id: 1674149045316657511
  name: main2
spec:
  enabled: true
  role_map:
  - local:
    - arc
    remote: '*'
  token: 0a9a0bb03ce2f801a197754e086fc69b
  tunnel_addr: tel11.arctevity-cloud.com:3024
  web_proxy_addr: tel11.arctevity-cloud.com:3080
version: v2
arc@remote1:/etc $
```

However, same offline result as before:

```
ubuntu@tel11:/var/lib/teleport$ sudo tsh clusters --proxy=tel11.arctevity-cloud.com --user=alan.lubold@arctevity.com
Enter password for Teleport user alan.lubold@arctevity.com:
Enter your OTP token:
Cluster Name Status  Cluster Type Labels Selected 
------------ ------- ------------ ------ -------- 
main2        online  root                *        
8.1.1.2.1    offline leaf                         
test11       offline leaf                         

ubuntu@tel11:/var/lib/teleport$
```

Log entries from root server:

```
{"ei":2147483647,"event":"session.upload","uid":"1df19d15-8442-46a0-954a-91cb4555c94f","code":"T2005I","time":"2023-01-19T14:08:46.706Z","cluster_name":"main2","sid":"6e18b649-d7c2-4440-90a7-37aec0e5e35f","url":"file:///var/lib/teleport/log/records/multi/6e18b649-d7c2-4440-90a7-37aec0e5e35f"}
{"ei":0,"event":"trusted_cluster_token.create","uid":"e90d834a-18b6-4ce6-9e82-943b1f63ac08","code":"T7002I","time":"2023-01-19T17:22:12.359Z","cluster_name":"main2","expires":"0001-01-01T00:00:00Z","user":"109f33cc-3243-4e2e-8062-4eff5847d11d.main2"}
{"ei":0,"event":"user.login","uid":"9a01bfc3-994b-4237-a599-8d19359c8acc","code":"T1000I","time":"2023-01-19T17:25:21.286Z","cluster_name":"main2","user":"alan.lubold@arctevity.com","success":true,"method":"local","mfa_device":{"mfa_device_name":"otp-device","mfa_device_uuid":"871cb4aa-9983-4ca0-99ed-8b130184f184","mfa_device_type":"TOTP"},"user_agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36","addr.remote":"73.149.126.150:51767"}
{"ei":0,"event":"cert.create","uid":"13ad580f-8031-477a-847a-7fc186ccf824","code":"TC000I","time":"2023-01-19T17:25:21.311Z","cluster_name":"main2","cert_type":"user","identity":{"user":"alan.lubold@arctevity.com","roles":["admin"],"logins":["ubuntu","arc","-teleport-internal-join"],"expires":"2023-01-20T05:25:21.30416827Z","route_to_cluster":"main2","traits":{"aws_role_arns":null,"db_names":null,"db_users":null,"kubernetes_groups":null,"kubernetes_users":null,"logins":null,"windows_logins":null},"teleport_cluster":"main2","prev_identity_expires":"0001-01-01T00:00:00Z"}}
```

Do I have a role problem?

Thanks

**Reply by webvictim** [reply]
I don't think you have a role problem - I think you have a connectivity problem.

I see from your original file that your root cluster is using `proxy_listener_mode: multiplex`, so your trusted cluster config should in fact be using the web port to establish its reverse tunnel (not port 3024). Something like this:
```yaml
kind: trusted_cluster
metadata:
  name: main2
spec:
  enabled: true
  role_map:
  - local:
    - arc
    remote: '*'
  token: <token>
  tunnel_addr: tel11.arctevity-cloud.com:3080
  web_proxy_addr: tel11.arctevity-cloud.com:3080
```

(you'll need to delete the existing trusted cluster config and re-add it with the changed config and a valid token)

**Comment by alanlubold** [comment]
Here are the roles, in case that is useful, in the attached to preserve formatting.

[roles.txt](https://github.com/gravitational/teleport/files/10459675/roles.txt)

**Comment by alanlubold** [comment]
Here is the result of **sudo journalctl -fu teleport | more** on the root server, while logging into the web interface and then trying to select the leaf cluster **test11**.  The error message on the web interface is **cluster "test11" is not found**

[root_log.txt](https://github.com/gravitational/teleport/files/10468520/root_log.txt)

**Comment by alanlubold** [comment]
Forgot to mention, using the web interface (or the CLI) works fine to establish an SSH connection to the root server.

**Comment by alanlubold** [comment]
Gus,

 

Thank you, this looks promising and I’ll try it now.

 

Alan

 

From: Gus Luxton ***@***.***>
Reply-To: gravitational/teleport ***@***.***>
Date: Friday, January 20, 2023 at 3:09 PM
To: gravitational/teleport ***@***.***>
Cc: alanlubold ***@***.***>, Author ***@***.***>
Subject: Re: [gravitational/teleport] failure to join a trusted cluster (Discussion #20301)

 

I don't think you have a role problem - I think you have a connectivity problem.

I see from your original file that your root cluster is using proxy_listener_mode: multiplex, so your trusted cluster config should in fact be using the web port to establish its reverse tunnel (not port 3024). Something like this:

kind: trusted_cluster
metadata:
  name: main2
spec:
  enabled: true
  role_map:
  - local:
    - arc
    remote: '*'
  token: <token>
  tunnel_addr: tel11.arctevity-cloud.com:3080
  web_proxy_addr: tel11.arctevity-cloud.com:3080
(you'll need to delete the existing trusted cluster config and re-add it with the changed config and a valid token)

**Comment by alanlubold** [comment, marked as answer]
That worked, thank you very much for your patience and help.

Alan

**Reply by webvictim** [reply]
No problem, glad we got there!
